xds/internal/balancer/ringhash/picker.go

```diff
@@ -0,0 +1,140 @@
+/*
+ *
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package ringhash
+
+import (
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/status"
+)
+
+type picker struct {
+	ring *ring
+}
+
+func newPicker(ring *ring) *picker {
+	return &picker{ring: ring}
+}
+
+// handleRICS generates pick result if the entry is in Ready, IDLE, Connecting
+// or Shutdown. TransientFailure will be handled specifically after this
+// function returns.
+//
+// The first return value indicates if the state is in Ready, IDLE, Connecting
+// or Shutdown. If it's true, the PickResult and error should be returned from
+// Pick() as is.
+func handleRICS(e *ringEntry) (bool, balancer.PickResult, error) {
```

Comment:
Nit: I believe the `ok` return value should be last, since that determines whether the other values are valid, and usually comes last in Go APIs.

Reply:
Done.
I thought lint would complain if `error` is not the last return value.

Reply:
Lint did complain.
```
+ golint ./...
xds/internal/balancer/ringhash/picker.go:45:1: error should be the last type when returning multiple items
```

I reverted the commit.

Reply:
Semantically that's wrong, though.  `error` here is a value, not a result, and the `bool` is part of the result.

A couple options: `type pickError = error` and return `pickError` instead, or put `PickResult` and `error` inside a struct?

Reply:
Added a struct